core/src/main/java/org/infinispan/iteration/impl/TransactionAwareCloseableIterator.java

```diff
@@ -19,45 +26,71 @@
+         KeyValueFilter<? super K, ? super V> filter,
+         Converter<? super K, ? super V, ? extends C> converter, 
+         TxInvocationContext<LocalTransaction> ctx, Cache<K, V> cache) {
       super(realIterator, cache, false);
       this.ctx = ctx;
+      this.filter = filter;
+      this.converter = checkForKeyValueFilterConverter(filter, converter);
+      this.entryFactory = cache.getAdvancedCache().getComponentRegistry().getComponent(
+            InternalEntryFactory.class);
       contextEntries = new ArrayList<>(ctx.getLookedUpEntries().values());
       currentValue = getNextFromIterator();
    }
 
+   protected CacheEntry<K, V> filterEntry(CacheEntry<K, V> entry) {
+      if (converter == null && filter instanceof KeyValueFilterConverter) {
+         K key = entry.getKey();
+         C converted = ((KeyValueFilterConverter<K, V, C>)filter).filterAndConvert(
+               key, entry.getValue(), entry.getMetadata());
+         if (converted != null) {
+            entry = entryFactory.create(entry);
+            entry.setValue((V) converted);
+            return entry;
+         }
+      } else if (filter == null || 
+            filter.accept(entry.getKey(), entry.getValue(), entry.getMetadata())) {
+         return entry;
+      }
+      return null;
+   }
+
    @Override
    protected CacheEntry<K, C> getNextFromIterator() {
-      CacheEntry<K, C> returnedEntry = null;
+      CacheEntry<K, V> returnedEntry = null;
       // We first have to exhaust all of our context entries
-      CacheEntry<K, C> entry;
-      while (!contextEntries.isEmpty() && (entry = contextEntries.remove(0)) != null) {
+      CacheEntry<K, V> entry;
+      while (returnedEntry == null && !contextEntries.isEmpty() && 
+            (entry = contextEntries.remove(0)) != null) {
          seenContextKeys.add(entry.getKey());
          if (!ctx.isEntryRemovedInContext(entry.getKey())) {
-            returnedEntry = entry;
+            returnedEntry = filterEntry(entry);
          }
       }
       if (returnedEntry == null) {
          while (realIterator.hasNext()) {
-            CacheEntry<K, C> iteratedEntry = realIterator.next();
+            CacheEntry<K, V> iteratedEntry = realIterator.next();
             CacheEntry contextEntry;
             // If the value was in the context then we ignore the stored value since we use the context value
             if ((contextEntry = ctx.lookupEntry(iteratedEntry.getKey())) != null) {
-               if (seenContextKeys.add(contextEntry.getKey()) && !contextEntry.isRemoved()) {
-                  returnedEntry = contextEntry;
+               if (seenContextKeys.add(contextEntry.getKey()) && !contextEntry.isRemoved() && 
+                     (returnedEntry = filterEntry(contextEntry)) != null) {
                   break;
                }
                
-            } else {
-               returnedEntry = iteratedEntry;
+            } else if ((returnedEntry = filterEntry(iteratedEntry)) != null) {
```

Comment:
I've just realized this change is not needed because it causes already filtered (and possibly converted) entries to be re-filtered. If they were already converted then the new type might no longer be suitable for re-filtering and will not pass the filter or will cause CCE. If they were not converted then the re-filtering is harmless but unnecessary.